x/metadata/keeper/query_server.go

```diff
@@ -358,7 +356,7 @@ func (k Keeper) SessionsAll(c context.Context, req *types.SessionsAllRequest) (*
 		return nil // Still want to move on to the next.
 	})
 	if err != nil {
-		return &retval, status.Error(codes.Unavailable, err.Error())
+		return &retval, sdkerrors.ErrJSONUnmarshal.Wrap(err.Error())
```

Comment:
This is tricky. I think I understand why you have ErrJSONUnmarshal, but it looks like it could also be InvalidRequest. 

```
if offset > 0 && key != nil {
		return nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
	}
```

I'm not entirely sure which one is correct for all the Paginate errors, but my guess would be InvalidRequest since it's more generic.

Reply:
After reviewing a little deeper.  I think they should be InvalidRequest errors too.  The  onResult functions don't return an error on Unmarshalling errors.  Which leaves the only possible error that can return from the Pagination function is the `fmt.Errorf("invalid request, either offset or key is expected, got both")`.